web/src/connectors/task-annotator-connector/task-annotator-context.tsx

```diff
@@ -316,77 +350,87 @@ export const TaskAnnotatorContextProvider: React.FC<ProviderProps> = ({
+        );
+    }
+
+    const isDocumentExist = !!documentsResult.data;
+    const isSplitValidation = task?.is_validation && job?.validation_type === 'extensive_coverage';
+    const isOnePDFLoadingFlow =
+        !isSplitValidation && !selectedRelatedDoc && fileMetaInfo.extension === '.pdf';
+
+    const documentDataViaLazyLoading = useDocumentDataLazyLoading(
         {
-            jobId: getJobId(),
-            fileId: getFileId(),
+            task,
+            job,
+            jobId,
+            fileMetaInfo,
             revisionId,
             pageNumbers,
-            userId:
-                job?.validation_type === 'extensive_coverage' && !revisionId ? task?.user_id : ''
+            setSelectedAnnotation
         },
-        { enabled: Boolean(task || job) }
+        { enabled: isDocumentExist && isOnePDFLoadingFlow }
     );
 
-    const tokenRes = useTokens(
+    const documentDataViaFullLoading = useDocumentDataFullLoading(
         {
-            fileId: getFileId(),
-            pageNumbers: pageNumbers.slice(tokenPagesChunk, tokenPagesChunk + PageTokensChunkSize)
+            task,
+            job,
+            jobId,
+            fileMetaInfo,
+            revisionId,
+            pageNumbers
         },
-        { enabled: tokenPagesChunk < pageNumbers.length }
+        { enabled: isDocumentExist && !isOnePDFLoadingFlow }
     );
 
-    useEffect(() => {
-        if (tokenRes.status === 'success') {
-            if (tokenRes.data && tokenRes.data.length && tokenPagesChunk < pageNumbers.length) {
-                let newTokensPages = [];
-
-                tokenRes.data.forEach((token) => {
-                    const thisTokenPage = tokensPagesState.find(
-                        (tokenPage: PageInfo) => tokenPage.page_num === token.page_num
-                    );
-
-                    if (!thisTokenPage) newTokensPages.push(token);
-
-                    setTokensPagesState([...tokensPagesState, ...tokenRes.data]);
-                    setTokenPagesChunk(tokenPagesChunk + PageTokensChunkSize);
-                });
-            }
-        }
-    }, [tokenRes]);
-
-    const tokenPages = tokensPagesState;
-
-    if (!fileMetaInfo) {
-        fileMetaInfo = useMemo(
-            () => ({
-                ...documentSearchResultMapper(documentsResult.data),
-                isLoading: isTaskLoading || documentsResult.isLoading
-            }),
-            [documentsResult.data, documentsResult.isLoading, isTaskLoading]
-        );
+    let documentData;
+
+    // TODO: This is a workaround to use different loading approaches in one big context.
+    // For PDF and other type of doc different context should be used.
+    // This logic should be moved in appropriate places where it's needed (this
+    // requires refactoring).
+    if (!isDocumentExist) {
+        documentData = defaultDocumentData;
+    } else {
+        documentData = isOnePDFLoadingFlow
```

Comment:
I've moved the previous logic where we load all data in the separate hook, the same about the new logic. The logic in inside `documentDataViaFullLoading` is just a movement of the code which we had before.`documentDataViaLazyLoading` contains new logic about lazy loading.